### PR TITLE
Add limit pushdown rule

### PR DESCRIPTION
**Pull request by FX196**
This PR adds a custom catalyst rule for doing limit pushdown for Delta Sharing queries. When the rule is applied, the limit in SQL queries will be pushed down and sent to the server as `limitHints`. e.g. `SELECT * FROM table LIMIT 100`

Added tests.

### CODE REVIEW

File: spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSuite.scala [hunk 1]
```diff
@@ -22,7 +22,7 @@ import scala.util.Random
 
 import org.apache.commons.io.IOUtils
 import org.apache.hadoop.fs.{FileSystem, Path}
-import org.apache.spark.sql.{QueryTest, Row}
+import org.apache.spark.sql.{QueryTest, Row, SparkSession}
```
Comment (zsxwing) [inline spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSuite.scala:25]: nit: unused change

File: spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala [hunk 2]
```diff
@@ -55,7 +56,11 @@ private[sharing] class RemoteDeltaLog(
   def createRelation(): BaseRelation = {
     val spark = SparkSession.active
     val snapshotToUse = snapshot
-    val fileIndex = new RemoteDeltaFileIndex(spark, this, path, snapshotToUse)
+    val fileIndex = new RemoteDeltaFileIndex(spark, this, path, snapshotToUse, None)
+    if (spark.sessionState.conf.getConfString(
+      "spark.delta.sharing.limitPushdown.enabled", "false").toBoolean) {
```
Comment (zsxwing) [inline spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala:61]: this should be enabled by default

File: spark/src/main/scala/io/delta/sharing/spark/perf/DeltaSharingLimitPushDown.scala [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+object DeltaSharingLimitPushDown extends Rule[LogicalPlan] {
+
+  def setup(spark: SparkSession): Unit = synchronized {
+    if (!spark.experimental.extraOptimizations.contains(DeltaSharingLimitPushDown) ) {
+      spark.experimental.extraOptimizations ++= Seq(DeltaSharingLimitPushDown)
+    }
+  }
+
+  def apply(p: LogicalPlan): LogicalPlan = {
+    p transform {
```
Comment (zsxwing) [inline spark/src/main/scala/io/delta/sharing/spark/perf/DeltaSharingLimitPushDown.scala:36]: let's also check `plan.conf.getConfString("spark.delta.sharing.limitPushdown.enabled", "true")`. If it's not enabled, we just return `p` directly.

File: python/delta_sharing/version.py [hunk 4]
```diff
@@ -14,4 +14,4 @@
 # limitations under the License.
 #
 
-__version__ = "0.3.0.dev0"
+__version__ = ""
```
Comment (zsxwing) [inline python/delta_sharing/version.py:17]: unnecessary change

**Review by zsxwing — COMMENTED** [review]
One minor comment. Otherwise, LGTM

File: python/delta_sharing/version.py (hunk 4 above)
Comment (FX196) [reply python/delta_sharing/version.py]: Sorry don't know how that got in there. Reverted.

**Review by zsxwing — APPROVED** [review]
